internal/controllers/app_controller.go

```diff
@@ -210,8 +207,38 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App) reconci
 
 	// check for canary deployment
 	if app.Spec.Canary.Active {
+		// ensures that the canary deployment exists
+		if len(app.Spec.Deployments) <= 1 {
+			app.Spec.Canary.Active = false
+			return reconcileResult{
+				status:     v1.ConditionFalse,
+				message:    "no canary deployment found",
+				useTimeout: true,
+			}
+		}
+
 		// retry until all pods for canary deployment comes to running state.
 		if err := checkPodStatus(r.Client, app.Name, app.Spec.Deployments[1].Version); err != nil {
+			// increment failure count for canary
+			app.IncrementCanaryFailureCounter()
+			if e := r.Update(ctx, app); err != nil {
+				return reconcileResult{
+					status:     v1.ConditionFalse,
+					message:    fmt.Sprintf("failed to increment canary failure count: %v", e),
+					useTimeout: true,
+				}
+			}
+
+			// check if rollback is needed
+			app.DoRollbackIfNeeded()
```

Comment:
can we have only one Update operation or update the second time only if we have to update the app. 

```golang

app.IncrementFailureCount()
app.DoRollbackIfNeeded() 
r.update()..
```